File: packages/engine/src/cores/famibob/apu/apu.js
import { Region } from '..'
import {
  Mixer,
  FrameCounter,
  Square,
  Triangle,
  Noise,
  DMC,
  SQUARE_CHANNEL,
  FRAME_TYPE,
  DMC_LOOKUP_TABLE_NTSC,
  DMC_LOOKUP_TABLE_PAL,
  NOISE_LOOKUP_TABLE_NTSC,
  NOISE_LOOKUP_TABLE_PAL,
  STEP_CYCLES_NTSC,
  STEP_CYCLES_PAL,
} from '.'

export class APU {
  constructor(region, mixer) {
    this.enabled = true
    this.region = region
    this.mixer = mixer
    this.currentCycle = 0
    this.previousCycle = 0
    this.needsToRun = false
    this.frameCounter = new FrameCounter(this)
    this.square1 = new Square(mixer, SQUARE_CHANNEL.ONE)
    this.square2 = new Square(mixer, SQUARE_CHANNEL.TWO)
    this.triangle = new Triangle(mixer)
    this.noise = new Noise(mixer)
    this.dmc = new DMC(mixer)
  }

  read(address) {
    this.run()

    const v =
      0 |
      (this.square1.status ? 0x01 : 0) |
      (this.square2.status ? 0x02 : 0) |
      (this.triangle.status ? 0x04 : 0) |
      (this.noise.status ? 0x08 : 0) |
      (this.dmc.status ? 0x10 : 0) |
      (this.frameCounter.irq ? 0x40 : 0) |
      (this.dmc.irq ? 0x80 : 0)

    this.frameCounter.irq = false
    return v
  }

  write(address, value) {
    this.run()
    switch (address) {
      case 0x4000: {
        this.square1.envelope.initEnvelope(value)
        this.needsToRun = true
        this.square1.duty = (value & 0xc0) >>> 6
        if (!this.square1.isMMC5Square) this.square1.updateOutput()
        break
      }
      case 0x4001: {
        this.square1.initSweep(value)
        if (!this.square1.isMMC5Square) this.square1.updateOutput()
        break
      }
      case 0x4002: {
        this.square1.setPeriod((this.square1.realPeriod & 0x0700) | value)
        if (!this.square1.isMMC5Square) this.square1.updateOutput()
        break
      }
      case 0x4003: {
        this.square1.envelope.lengthCounter.loadLengthCounter(value >>> 3)
        this.needsToRun = true
        this.square1.setPeriod((this.square1.realPeriod & 0xff) | ((value & 0x07) << 8))
        this.square1.dutyPos = 0
        this.square1.envelope.resetEnvelope()
        if (!this.square1.isMMC5Square) this.square1.updateOutput()
        break
      }
      case 0x4004: {
        this.square2.envelope.initEnvelope(value)
        this.needsToRun = true
        this.square2.duty = (value & 0xc0) >>> 6
        if (!this.square2.isMMC5Square) this.square2.updateOutput()
        break
      }
      case 0x4005: {
        this.square2.initSweep(value)
        if (!this.square2.isMMC5Square) this.square2.updateOutput()
        break
      }
      case 0x4006: {
        this.square2.setPeriod((this.square2.realPeriod & 0x0700) | value)
        if (!this.square2.isMMC5Square) this.square2.updateOutput()
        break
      }
      case 0x4007: {
        this.square2.envelope.lengthCounter.loadLengthCounter(value >>> 3)
        this.needsToRun = true
        this.square2.setPeriod((this.square2.realPeriod & 0xff) | ((value & 0x07) << 8))
        this.square2.dutyPos = 0
        this.square2.envelope.resetEnvelope()
        if (!this.square2.isMMC5Square) this.square2.updateOutput()
        break
      }
      case 0x4008: {
        this.triangle.linearControlFlag = (value & 0x80) === 0x80
        this.triangle.linearCounterReload = value & 0x7f
        this.triangle.lengthCounter.initLengthCounter(this.triangle.linearControlFlag)
        this.needsToRun = true
        break
      }
      case 0x400a: {
        this.triangle.timer.period = (this.triangle.timer.period & ~0x00ff) | value
        break
      }
      case 0x400b: {
        this.triangle.lengthCounter.loadLengthCounter(value >>> 3)
        this.needsToRun = true
        this.triangle.timer.period = (this.triangle.timer.period & ~0xff00) | ((value & 0x07) << 8)
        this.triangle.linearReloadFlag = true
        break
      }
      case 0x400c: {
        this.noise.envelope.initEnvelope(value)
        this.needsToRun = true
        break
      }
      case 0x400e: {
        this.noise.timer.period = this.noise.lookupTable[value & 0x0f] - 1
        this.noise.modeFlag = (value & 0x80) === 0x80
        break
      }
      case 0x400f: {
        this.noise.envelope.lengthCounter.loadLengthCounter(value >>> 3)
        this.needsToRun = true
        this.noise.envelope.resetEnvelope()
        break
      }
      case 0x4010: {
        this.dmc.irqEnabled = (value & 0x80) === 0x80
        this.dmc.loopFlag = (value & 0x40) === 0x40
        this.dmc.timer.period = this.dmc.lookupTable[value & 0x0f] - 1
        if (!this.dmc.irqEnabled) this.dmc.irq = false
        break
      }
      case 0x4011: {
        this.dmc.outputLevel = value & 0x7f
        this.dmc.timer.addOutput(this.dmc.outputLevel)
        break
      }
      case 0x4012: {
        this.dmc.sampleAddr = 0xc000 | (value << 6)
        break
      }
      case 0x4013: {
        this.dmc.sampleLength = (value << 4) | 0x0001
        break
      }
      case 0x4015: {
        this.dmc.irq = false
        this.square1.setEnabled((value & 0x01) > 0)
        this.square2.setEnabled((value & 0x02) > 0)
        this.triangle.setEnabled((value & 0x04) > 0)
        this.noise.setEnabled((value & 0x08) > 0)
        this.dmc.setEnabled((value & 0x10) > 0)
        break
      }
      case 0x4017: {
        this.frameCounter.newValue = value
        this.frameCounter.writeDelayCounter = 3
        this.frameCounter.inhibitIrq = (value & 0x40) === 0x40
        if (this.frameCounter.inhibitIrq) {
          this.frameCounter.irq = false
        }
        break
      }
    }
  }

  process() {
    if (this.enabled) {
      this.currentCycle++
      if (this.currentCycle === this.mixer.cycleLength - 1) {
        this.endFrame()
      } else if (this.itNeedsToRun(this.currentCycle)) {
        this.run()
      }
    }
  }

  setRegion(region) {
    this.region = region
    this.mixer.setRegion(region)
    this.dmc.lookupTable = region === Region.NTSC ? DMC_LOOKUP_TABLE_NTSC : DMC_LOOKUP_TABLE_PAL
    this.noise.lookupTable = region === Region.NTSC ? NOISE_LOOKUP_TABLE_NTSC : NOISE_LOOKUP_TABLE_PAL
    this.run()
    this.frameCounter.stepCycles = region === Region.NTSC ? STEP_CYCLES_NTSC : STEP_CYCLES_PAL
  }

  endFrame() {
    this.run()
    this.square1.endFrame()
    this.square2.endFrame()
    this.triangle.endFrame()
    this.noise.endFrame()
    this.dmc.endFrame()
    this.mixer.playAudioBuffer(this.currentCycle)
    this.currentCycle = 0
    this.previousCycle = 0
  }

  itNeedsToRun(cycle) {
    if (this.dmc.itNeedsToRun() || this.needsToRun) {
      this.needsToRun = false
      return true
    } else {
      const cycleToRun = cycle - this.previousCycle
      return this.frameCounter.itNeedsToRun(cycleToRun) || this.dmc.irqPending(cycleToRun)
    }
  }

  run() {
    let cyclesToRun = this.currentCycle - this.previousCycle
    while (cyclesToRun > 0) {
      const ref = { cyclesToRun }
      this.previousCycle += this.frameCounter.run(ref)
      cyclesToRun = ref.cyclesToRun

      this.square1.reloadLengthCounter()
      this.square2.reloadLengthCounter()
      this.noise.reloadLengthCounter()
      this.triangle.reloadLengthCounter()
      this.square1.run(this.previousCycle)
      this.square2.run(this.previousCycle)
      this.noise.run(this.previousCycle)
      this.triangle.run(this.previousCycle)
      this.dmc.run(this.previousCycle)
    }
  }

  frameCounterTick(frameType) {
    this.square1.tickEnvelope()
    this.square2.tickEnvelope()
    this.triangle.tickLinearCounter()
    this.noise.tickEnvelope()
    if (frameType === FRAME_TYPE.HALF_FRAME) {
      this.square1.tickLengthCounter()
      this.square2.tickLengthCounter()
      this.triangle.tickLengthCounter()
      this.noise.tickLengthCounter()
      this.square1.tickSweep()
      this.square2.tickSweep()
    }
  }

  reset() {
    this.enabled = true
    this.currentCycle = 0
    this.previousCycle = 0
    this.needsToRun = false
    this.square1.reset()
    this.square2.reset()
    this.triangle.reset()
    this.noise.reset()
    this.dmc.reset()
    this.frameCounter.reset()
    this.mixer.reset()
  }
}

File: packages/engine/src/cores/famibob/apu/apu.zig
const std = @import("std");
const Region = @import("../../core.zig").Region;
const Memory = @import("../../../memory.zig").Memory;

pub const LengthCounter = @import("length_counter.zig").LengthCounter;
pub const FrameCounter = @import("frame_counter.zig").FrameCounter;
pub const AudioChannel = @import("audio_channel.zig").AudioChannel;
pub const Envelope = @import("envelope.zig").Envelope;
pub const Triangle = @import("triangle.zig").Triangle;
pub const Square = @import("square.zig").Square;
pub const Noise = @import("noise.zig").Noise;
pub const Timer = @import("timer.zig").Timer;
pub const Mixer = @import("mixer.zig").Mixer;
pub const DMC = @import("dmc.zig").DMC;

pub const APU = struct {
    allocator: std.mem.Allocator,
    enabled: bool = true,
    mixer: *Mixer,
    region: Region,

    current_cycle: u32 = 0,
    previous_cycle: u32 = 0,
    needs_to_run: bool = false,

    frame_counter: FrameCounter = .{},
    square1: Square(.one) = .{},
    square2: Square(.two) = .{},
    triangle: Triangle = .{},
    noise: Noise = .{},
    dmc: DMC = .{},

    pub fn init(allocator: std.mem.Allocator, region: Region, mixer: *Mixer) !*APU {
        const instance = try allocator.create(APU);

        instance.* = .{
            .allocator = allocator,
            .region = region,
            .mixer = mixer,
        };

        instance.frame_counter.apu = instance;
        instance.square1.timer.mixer = mixer;
        instance.square2.timer.mixer = mixer;
        instance.triangle.timer.mixer = mixer;
        instance.noise.timer.mixer = mixer;
        instance.dmc.timer.mixer = mixer;

        instance.reset();
        return instance;
    }

    pub fn deinit(self: *APU) void {
        self.allocator.destroy(self);
    }

    fn deinitMemory(ctx: *anyopaque) void {
        const self: *@This() = @ptrCast(@alignCast(ctx));
        self.deinit();
    }

    pub fn read(ctx: *anyopaque, address: u16) u8 {
        const self: *@This() = @ptrCast(@alignCast(ctx));
        _ = address;

        self.run();

        const v = 0 //
        | @as(u8, if (self.square1.getStatus()) 0x01 else 0) //
        | @as(u8, if (self.square2.getStatus()) 0x02 else 0) //
        | @as(u8, if (self.triangle.getStatus()) 0x04 else 0) //
        | @as(u8, if (self.noise.getStatus()) 0x08 else 0) //
        | @as(u8, if (self.dmc.getStatus()) 0x10 else 0) //
        | @as(u8, if (self.frame_counter.irq_requested) 0x40 else 0) //
        | @as(u8, if (self.dmc.irq_requested) 0x80 else 0);

        self.frame_counter.irq_requested = false;
        return v;
    }

    pub fn write(ctx: *anyopaque, address: u16, value: u8) void {
        const self: *@This() = @ptrCast(@alignCast(ctx));
        self.run();

        switch (address) {
            0x4000 => {
                self.square1.envelope.initEnvelope(value);
                self.needs_to_run = true;
                self.square1.duty = (value & 0xC0) >> 6;
                if (!self.square1.is_mmc5_square) self.square1.updateOutput();
            },
            0x4001 => {
                self.square1.initSweep(value);
                if (!self.square1.is_mmc5_square) self.square1.updateOutput();
            },
            0x4002 => {
                self.square1.setPeriod((self.square1.real_period & 0x0700) | value);
                if (!self.square1.is_mmc5_square) self.square1.updateOutput();
            },
            0x4003 => {
                self.square1.envelope.length_counter.loadLengthCounter(value >> 3);
                self.needs_to_run = true;
                self.square1.setPeriod((self.square1.real_period & 0xFF) | (@as(u16, value & 0x07) << 8));
                self.square1.duty_pos = 0;
                self.square1.envelope.resetEnvelope();
                if (!self.square1.is_mmc5_square) self.square1.updateOutput();
            },
            0x4004 => {
                self.square2.envelope.initEnvelope(value);
                self.needs_to_run = true;
                self.square2.duty = (value & 0xC0) >> 6;
                if (!self.square2.is_mmc5_square) self.square2.updateOutput();
            },
            0x4005 => {
                self.square2.initSweep(value);
                if (!self.square2.is_mmc5_square) self.square2.updateOutput();
            },
            0x4006 => {
                self.square2.setPeriod((self.square2.real_period & 0x0700) | value);
                if (!self.square2.is_mmc5_square) self.square2.updateOutput();
            },
            0x4007 => {
                self.square2.envelope.length_counter.loadLengthCounter(value >> 3);
                self.needs_to_run = true;
                self.square2.setPeriod((self.square2.real_period & 0xFF) | (@as(u16, value & 0x07) << 8));
                self.square2.duty_pos = 0;
                self.square2.envelope.resetEnvelope();
                if (!self.square2.is_mmc5_square) self.square2.updateOutput();
            },
            0x4008 => {
                self.triangle.linear_control_flag = (value & 0x80) == 0x80;
                self.triangle.linear_counter_reload = value & 0x7F;
                self.triangle.length_counter.initLengthCounter(self.triangle.linear_control_flag);
                self.needs_to_run = true;
            },
            0x400A => {
                self.triangle.timer.period = (self.triangle.timer.period & 0x700) | value;
            },
            0x400B => {
                self.triangle.length_counter.loadLengthCounter(value >> 3);
                self.needs_to_run = true;
                self.triangle.timer.period = (self.triangle.timer.period & 0xFF) | (@as(u16, value & 0x07) << 8);
                self.triangle.linear_reload_flag = true;
            },
            0x400C => {
                self.noise.envelope.initEnvelope(value);
                self.needs_to_run = true;
            },
            0x400E => {
                self.noise.timer.period = self.noise.lookup_table[value & 0x0F] - 1;
                self.noise.mode_flag = (value & 0x80) == 0x80;
            },
            0x400F => {
                self.noise.envelope.length_counter.loadLengthCounter(value >> 3);
                self.needs_to_run = true;
                self.noise.envelope.resetEnvelope();
            },
            0x4010 => {
                self.dmc.irq_enabled = (value & 0x80) == 0x80;
                self.dmc.loop_flag = (value & 0x40) == 0x40;
                self.dmc.timer.period = self.dmc.lookup_table[value & 0x0F] - 1;
                if (!self.dmc.irq_enabled) self.dmc.irq_requested = false;
            },
            0x4011 => {
                const new_value: i8 = @bitCast(value & 0x7f);
                const old_value: i8 = @bitCast(self.dmc.output_level);
                self.dmc.output_level = @bitCast(new_value);

                if (@abs(new_value - old_value) > 50) {
                    // reduce popping sounds
                    self.dmc.output_level = @bitCast(new_value - @divTrunc(new_value - old_value, 2));
                }
            },
            0x4012 => {
                self.dmc.sample_addr = 0xC000 | (@as(u16, value) << 6);
            },
            0x4013 => {
                self.dmc.sample_length = (@as(u16, value) << 4) | 0x0001;
            },
            0x4015 => {
                self.dmc.irq_requested = false;
                self.square1.setEnabled((value & 0x01) > 0);
                self.square2.setEnabled((value & 0x02) > 0);
                self.triangle.setEnabled((value & 0x04) > 0);
                self.noise.setEnabled((value & 0x08) > 0);
                self.dmc.setEnabled((value & 0x10) > 0);
            },
            0x4017 => {
                self.frame_counter.new_value = value;
                self.frame_counter.write_delay_counter = 3;
                self.frame_counter.inhibit_irq = (value & 0x40) == 0x40;
                if (self.frame_counter.inhibit_irq) {
                    self.frame_counter.irq_requested = false;
                }
            },
            else => {},
        }
    }

    pub fn process(self: *APU) void {
        if (self.enabled) {
            self.current_cycle += 1;
            if (self.current_cycle == Mixer.cycle_length - 1) {
                self.endFrame();
            } else if (self.needsToRun(self.current_cycle)) {
                self.run();
            }
        }
    }

    pub fn setRegion(self: *APU, region: Region) void {
        self.region = region;
        self.mixer.setRegion(region);
        self.dmc.lookup_table = if (region == .ntsc) DMC.lookup_table_ntsc else DMC.lookup_table_pal;
        self.noise.lookup_table = if (region == .ntsc) Noise.lookup_table_ntsc else Noise.lookup_table_pal;
        self.run();
        self.frame_counter.step_cycles = if (region == .ntsc) FrameCounter.step_cycles_ntsc else FrameCounter.step_cycles_pal;
    }

    pub fn endFrame(self: *APU) void {
        self.run();
        self.square1.endFrame();
        self.square2.endFrame();
        self.triangle.endFrame();
        self.noise.endFrame();
        self.dmc.endFrame();
        self.mixer.endFrame(self.current_cycle);
        self.mixer.updateRates(false);
        self.current_cycle = 0;
        self.previous_cycle = 0;
    }

    fn needsToRun(self: *APU, cycle: u32) bool {
        if (self.dmc.needsToRun() or self.needs_to_run) {
            self.needs_to_run = false;
            return true;
        } else {
            const cycle_to_run: u32 = cycle - self.previous_cycle;
            return self.frame_counter.needsToRun(cycle_to_run) or self.dmc.irqPending(cycle_to_run);
        }
    }

    fn run(self: *APU) void {
        var cycles_to_run: i32 = @intCast(self.current_cycle - self.previous_cycle);
        while (cycles_to_run > 0) {
            self.previous_cycle += self.frame_counter.run(&cycles_to_run);
            self.square1.reloadLengthCounter();
            self.square2.reloadLengthCounter();
            self.noise.reloadLengthCounter();
            self.triangle.reloadLengthCounter();
            self.square1.run(self.previous_cycle);
            self.square2.run(self.previous_cycle);
            self.noise.run(self.previous_cycle);
            self.triangle.run(self.previous_cycle);
            self.dmc.run(self.previous_cycle);
        }
    }

    pub fn frameCounterTick(self: *APU, frame_type: FrameCounter.FrameType) void {
        self.square1.tickEnvelope();
        self.square2.tickEnvelope();
        self.triangle.tickLinearCounter();
        self.noise.tickEnvelope();
        if (frame_type == .half_frame) {
            self.square1.tickLengthCounter();
            self.square2.tickLengthCounter();
            self.triangle.tickLengthCounter();
            self.noise.tickLengthCounter();
            self.square1.tickSweep();
            self.square2.tickSweep();
        }
    }

    pub fn reset(self: *APU) void {
        self.enabled = true;
        self.current_cycle = 0;
        self.previous_cycle = 0;
        self.needs_to_run = false;
        self.square1.reset();
        self.square2.reset();
        self.triangle.reset();
        self.noise.reset();
        self.dmc.reset();
        self.frame_counter.reset();
        self.mixer.reset();
    }

    pub fn memory(self: *APU) Memory(u16, u8) {
        return .{
            .ptr = self,
            .vtable = &.{
                .read = read,
                .write = write,
                .deinit = deinitMemory,
            },
        };
    }

    pub fn jsonStringify(self: *const APU, jw: anytype) !void {
        try jw.beginObject();
        try jw.objectField("enabled");
        try jw.write(self.enabled);
        try jw.objectField("current_cycle");
        try jw.write(self.current_cycle);
        try jw.objectField("previous_cycle");
        try jw.write(self.previous_cycle);
        try jw.objectField("needs_to_run");
        try jw.write(self.needs_to_run);
        try jw.objectField("frame_counter");
        try jw.write(self.frame_counter);
        try jw.objectField("square1");
        try jw.write(self.square1);
        try jw.objectField("square2");
        try jw.write(self.square2);
        try jw.objectField("triangle");
        try jw.write(self.triangle);
        try jw.objectField("noise");
        try jw.write(self.noise);
        try jw.objectField("dmc");
        try jw.write(self.dmc);
        try jw.endObject();
    }
};
